Sparse tensor data structures

aestriplex opened the pull request:
This PR implements a first draft of the data structures required to implement sparse tensors support in ACL. 
Here are three points worth discussing:
1. In the COOTensor and CSRTensor classes, I implemented two different strategies for saving indexes. Specifically, in COO tensors, indexes are saved as `std::vector` (i.e., not in the same buffer provided by the tensor allocator, which is used by values instead). In CSR tensors, on the other hand, I stored them in the allocator buffer and saved only the corresponding (integer) offsets as class members. The method used in CSR is more efficient when working with the tensor (e.g., in a kernel/operator), so I thought I would extend it to the COO tensor as well. Are there any problems if this approach is too “low level”?
2. The test class should be converted into a Fixture.
3. As Pablo said in his comment on my design document in the mailing list, we need to make sure that the validation of current operators fails when using sparse tensors. A good idea, in my opinion, would be to create a “validate” method in `ICpuOperator` - and other kind of base operators -, so that it can be called within the validate of classes that implement this interface and do _not_ (yet) support sparse tensors (e.g., CpuGemmConv2d). This method, i.e., not performing the direct check `!src->info()->is_sparse()`, allows these checks to be extended in the future without having to modify all operators.

aestriplex committed:
sparse tensor data structures

COOTensor init
CSRTensor init
added SparseTensorAllocator
moved SparseTensor to arm_compute/core
fixed to_dense method for COOTensor
fixed to_sparse and to_dense methods for CSRTensor
changed CSR indices; print function moved to the bae class
implemented get_value to CSR and COO tensors
added SparseIterator

Change-Id: Id5ed095587662b750b89cbabd600a286ab9bf53b

morgolock commented on src/runtime/COOTensor.cpp:
```diff
@@ -0,0 +1,264 @@
+    _allocator.set_associated_memory_group(memory_group);
+}
+
+#ifdef ARM_COMPUTE_ASSERTS_ENABLED
+void COOTensor::print(std::ostream &os) const
+{
+    const uint8_t *data = static_cast<const uint8_t *>(buffer());
+
+    if(_indices.empty())
+    {
+        os << "index: [] values: []" << std::endl;
+        return;
+    }
+
+    for(size_t i = 0; i < _indices.size(); ++i)
+    {
+        const Coordinates &coord = _indices[i];
+        os << "index: [";
+        for (size_t j = 0; j < coord.num_dimensions(); ++j)
+        {
+            os << coord[j];
+            if (j < coord.num_dimensions() - 1) os << ", ";
+        }
+        os << "]  values: ";
+        print_values(os, data, i, dense_volume(sparse_dim()));
+    }
+}
+#endif // ARM_COMPUTE_ASSERTS_ENABLED
```
We already have a print method in https://github.com/ARM-software/ComputeLibrary/blob/main/src/core/ITensor.cpp

We should move this in `void ITensor::print(std::ostream &s, IOFormatInfo io_fmt) const` 

There you can ask

morgolock commented on src/runtime/CSRTensor.cpp:
```diff
@@ -0,0 +1,272 @@
+}
+
+#ifdef ARM_COMPUTE_ASSERTS_ENABLED
+void CSRTensor::print(std::ostream &os) const
+{
+    const uint8_t *_row_offsets = _allocator.data();
+    const uint8_t *_col_indices = _allocator.data() + _crow_bytes;
+    const uint8_t       *values = _allocator.data() + _crow_bytes + _col_bytes;
+
+    os << "r_offsets: [";
+    for(size_t i = 0; i < _crow_bytes / index_size; ++i)
+    {
+        os << *reinterpret_cast<const int32_t *>(_row_offsets + (i * index_size));
+        if (i < _crow_bytes / index_size - 1)
+        {
+            os << ", ";
+        }
+    }
+    os << "] cols: [";
+    for(size_t i = 0; i < _col_bytes / index_size; ++i)
+    {
+        os << *reinterpret_cast<const int32_t *>(_col_indices + (i * index_size));
+        if (i < _col_bytes / index_size - 1)
+        {
+            os << ", ";
+        }
+    }
+    os << "] values: ";
+    print_values(os, values, 0, nnz());
+}
+#endif // ARM_COMPUTE_ASSERTS_ENABLED
```
Move this to existing `void ITensor::print(std::ostream &s, IOFormatInfo io_fmt) const`

morgolock commented on tests/validation/cpu/unit/SparseTensor.cpp:
```diff
@@ -0,0 +1,467 @@
+TEST_SUITE(UNIT)
+TEST_SUITE(SparseTensor)
+
+TEST_CASE(ConvertDenseTensorToCOOTensor, framework::DatasetMode::ALL)
```
Yes, we'll need to rework the test using fixtures and helpers in ACL

I'd say lets first start with simple unit tests like:
- https://github.com/ARM-software/ComputeLibrary/blob/main/tests/validation/UNIT/TensorInfo.cpp
- https://github.com/ARM-software/ComputeLibrary/blob/main/tests/validation/NEON/UNIT/TensorAllocator.cpp

The simple tests should create the new sparse tensors and test the new methods.

Then more complex tests with operators will require fixtures:
- https://github.com/ARM-software/ComputeLibrary/blob/main/tests/validation/NEON/ArgMinMax.cpp
- https://github.com/ARM-software/ComputeLibrary/blob/main/tests/validation/fixtures/ArgMinMaxFixture.h

aestriplex replied:
I moved the conversion from/to sparse tensors tests in `validation/unit`, instead of `validation/cpu/unit`, and I used ACL helpers as done [https://github.com/ARM-software/ComputeLibrary/blob/main/tests/validation/UNIT/TensorInfo.cpp](https://github.com/ARM-software/ComputeLibrary/blob/main/tests/validation/UNIT/TensorInfo.cpp)

morgolock commented on src/runtime/SparseTensorAllocator.cpp:
```diff
@@ -0,0 +1,177 @@
+
+namespace arm_compute
+{
+SparseTensorAllocator::SparseTensorAllocator(IMemoryManageable *owner) : _owner(owner), _associated_memory_group(nullptr), _memory(), _values_bytes(0), _indices_bytes(0)
```
I think we should use the existing `TensorAllocator` to allocate any buffers.

morgolock reviewed (changes requested):
Hi @aestriplex 

I had a quick look at this and I think it's a good starting point. Please have a look and let me know your thoughts.

aestriplex committed:
cr fixes

Change-Id: I3a3fca2b842702672c915e1357602bf0fe47bd47

aestriplex commented on arm_compute/core/IReducibleTensor.h:
```diff
@@ -0,0 +1,76 @@
+
+#include "arm_compute/core/SparseTensor.h"
+#include "arm_compute/runtime/COOTensor.h"
+#include "arm_compute/runtime/CSRTensor.h"
```
I forward declared CSRTensor and COOTensor and removed the includes

aestriplex commented:
@morgolock Hi Pablo, this week I am going to continue implementing Transpose operator for both CSR and COO tensor. Are my fixes fine so far?
In particular, do you prefer indices to be represented as in CSRTensor (i.e., with allocators) or as in COOTensor (with C++ classes, i.e. std::vector)? Personally, I think the solution with allocators is by far the best in view of the implementation of operators. What do you think about it?

morgolock commented:
> Hi @morgolock Yes, I am willing to continue working on this, so I'm doing a rebase.
> 
> But I was waiting for a feedback for [#1186 (comment)](https://github.com/ARM-software/ComputeLibrary/pull/1186#discussion_r2408934953) and [#1186 (comment)](https://github.com/ARM-software/ComputeLibrary/pull/1186#discussion_r2408950345), to avoid a potential rollback on those fundamental implementations

Thanks @aestriplex , I'll have another look and provide feedback.

aestriplex committed:
Merge branch 'ARM-software:main' into sparse

aestriplex committed:
fixed CR comments

aestriplex committed:
CooTensor now uses the sparse allocator as CSRTensor

aestriplex committed:
added more tests for CSRTensor

aestriplex committed:
added support for NHWC COOTensor

aestriplex committed:
added support for NHWC CSRTensor

aestriplex commented:
Hi @morgolock I implemented the fixed for the copilot code review and added a few more commits  that implement:

1. More test cases
2. NHWC support for both CooTensor and CSRTensor

This completes the feature from a data structure point of view. Should I mark this PR as "ready for review" (i.e. exit from the "draft" stage)?

aestriplex commented:
> Hi @morgolock I implemented the fixed for the copilot code review and added a few more commits that implement:
> 
> 1. More test cases
> 2. NHWC support for both CooTensor and CSRTensor
> 
> This completes the feature from a data structure point of view. Should I mark this PR as "ready for review" (i.e. exit from the "draft" stage)?

@morgolock In addition to this, should I start implementing a Kernel for some basic operations, e.g. transpose?
